lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1219,14 +1219,119 @@ struct CSEVisitor : NodeWalker {
   }
 };
 
+/// A helper type for hashing Variable pointers when they are used as keys in
+/// hash maps for CSE. The hash is based on the type of the Variable (element
+/// type, dimensions), as well as a constant number of elements from the
+/// Variable to balance collisions with hash calclulation time.
+struct VarsHasherCSE {
+  size_t operator()(Variable *V) const {
+    auto hash = llvm::hash_value(V->getType());
+    auto &T = V->getPayload();
+    // Only use the first 8 elements in the hash. It's likely that if two
+    // tensors have different content they will diverge quickly. Fall back to
+    // full equality check in VarsEqCSE.
+    constexpr size_t maxNumEls = 8;
+    size_t numEls = std::min(T.getType().size(), maxNumEls);
+    size_t bufSize = T.getType().getElementSize() * numEls;
+    auto *data = T.getUnsafePtr();
+    for (size_t i = 0; i < bufSize; i++) {
+      hash = llvm::hash_combine(hash, data[i]);
+    }
+    return hash;
+  }
+};
+
+/// A helper type implementing the Variable equality predicate that can be used
+/// when Variable pointers are used as keys in hash maps for CSE. It is assumed
+/// the visibility and train kinds are the same, as CSE only inserts if Private
+/// and None, respectively.
+struct VarsEqCSE {
+  bool operator()(const Variable *lhs, const Variable *rhs) const {
+    // Only consider Vars for CSE if they have the same type. The train kind and
+    // visibility must already be the same.
+    if (lhs->getType() != rhs->getType()) {
+      return false;
+    }
+    assert(lhs->getVisibilityKind() == rhs->getVisibilityKind() &&
+           "Should only be comparing Variables with same VisibilityKind.");
+    assert(lhs->getTrainKind() == rhs->getTrainKind() &&
+           "Should only be comparing Variables with same TrainKind.");
+    // Only combine Vars if their data matches exactly, so allowed error is 0.0.
+    return lhs->getPayload().isEqual(rhs->getPayload(), /* allowedError */ 0.0);
+  }
+};
+
 } // namespace
 
+/// \returns true if Variable \p V is written into, either as a result, or as an
+/// overwritten input.
+static bool hasWriters(Variable *V) {
+  for (auto &U : V->getUsers()) {
+    auto *N = U.getUser();
+
+    // See if V is used as a result anywhere.
+    for (unsigned i = 0; i < N->getNumResults(); i++)
+      if (V == N->getNthResult(i))
+        return true;
+
+    // See if V is used as an overwritten input anywhere.
+    for (unsigned i = 0; i < N->getNumInputs(); i++)
+      if (N->isOverwrittenNthInput(i))
+        if (V == N->getNthInput(i))
+          return true;
+  }
+  return false;
+}
+
+/// Common Subexpression Elimination for Variables in the Module \p M.
+static void varsCSE(Module *M) {
```

Comment:
I think that a better name would be "deduplicate constants".

Reply:
Done.